src/components/implementation/no_interface/llbooter/boot_deps.h

```diff
@@ -94,73 +106,242 @@ boot_newcomp_sinv_alloc(spdid_t spdid)
+	cos_cap_cpy_at(new_comp_cap_info[spdid].compinfo, BOOT_CAPTBL_SELF_INITTHD_BASE,
+			boot_info, child_aep->thd);
+	cos_cap_cpy_at(new_comp_cap_info[spdid].compinfo, BOOT_CAPTBL_SELF_INITTCAP_BASE,
+			boot_info, child_aep->tc);
+	cos_cap_cpy_at(new_comp_cap_info[spdid].compinfo, BOOT_CAPTBL_SELF_INITRCV_BASE,
+			boot_info, child_aep->rcv);
+}
+
+static void
+boot_newcomp_create(spdid_t spdid, struct cos_compinfo *comp_info, int is_sched, int is_shdmem)
+{
+	compcap_t      cc;
+	captblcap_t    ct = new_comp_cap_info[spdid].compinfo->captbl_cap;
+	pgtblcap_t     pt = new_comp_cap_info[spdid].compinfo->pgtbl_cap;
+	sinvcap_t      sinv;
+	struct sl_thd *thd;
+
+	cc = cos_comp_alloc(boot_info, ct, pt, (vaddr_t)new_comp_cap_info[spdid].upcall_entry);
+
 	assert(cc);
-	new_comp_cap_info[spdid].compinfo->comp_cap = cc;
+	/* manually laying out struct due to upcall addr calculation */
+	boot_newcomp_definfo_init(spdid, cc, is_sched);
+
+	/* Create sinv capability from Components to Booter */
+	sinv = cos_sinv_alloc(boot_info, boot_info->comp_cap, (vaddr_t)__inv_test_entry,
+			(unsigned long)spdid);
 
-	/* Create sinv capability from Userspace to Booter components */
-	sinv = cos_sinv_alloc(&boot_info, boot_info.comp_cap, (vaddr_t)__inv_test_entry, token);
-	assert(sinv > 0);
+	assert(sinv);
 
-	cos_cap_cpy_at(new_comp_cap_info[spdid].compinfo, BOOT_CAPTBL_SINV_CAP, &boot_info, sinv);
+	cos_cap_cpy_at(new_comp_cap_info[spdid].compinfo, BOOT_CAPTBL_SINV_CAP, boot_info, sinv);
 
 	boot_newcomp_sinv_alloc(spdid);
 
-	main_thd = cos_initthd_alloc(&boot_info, cc);
-	assert(main_thd);
+	if (is_sched || is_shdmem) {
+		boot_newschedcomp_cap_init(spdid, ct, pt, cc);
 
-	/* Add created component to "scheduling" array */
-	while (schedule[i] != 0) i++;
-	schedule[i] = main_thd;
+		thd = sl_thd_comp_init(new_comp_cap_info[spdid].defcompinfo, is_sched);
+		assert(thd);
+		sl_thd_param_set(thd, sched_param_pack(SCHEDP_PRIO, THD_PRIO));
+	}
+
+	if (is_sched) {
+		sl_thd_param_set(thd, sched_param_pack(SCHEDP_BUDGET, THD_BUDGET * 1000));
+		sl_thd_param_set(thd, sched_param_pack(SCHEDP_WINDOW, THD_PERIOD * 1000));
+	}
 }
 
 static void
 boot_bootcomp_init(void)
 {
-	cos_meminfo_init(&boot_info.mi, BOOT_MEM_KM_BASE, COS_MEM_KERN_PA_SZ, BOOT_CAPTBL_SELF_UNTYPED_PT);
+	cos_defcompinfo_init();
+	boot_info = cos_compinfo_get(cos_defcompinfo_curr_get());
+	assert(boot_info);
 
-	cos_compinfo_init(&boot_info, BOOT_CAPTBL_SELF_PT, BOOT_CAPTBL_SELF_CT, BOOT_CAPTBL_SELF_COMP,
-	                  (vaddr_t)cos_get_heap_ptr(), BOOT_CAPTBL_FREE, &boot_info);
+	cos_meminfo_init(&(boot_info->mi), BOOT_MEM_KM_BASE, COS_MEM_KERN_PA_SZ,
+			BOOT_CAPTBL_SELF_UNTYPED_PT);
+	sl_init(SL_MIN_PERIOD_US);
 }
 
 static void
 boot_done(void)
 {
-	printc("Booter: done creating system.\n\n");
-	cos_thd_switch(schedule[sched_cur]);
+	printc("Booter: done creating system.\n");
+	printc("Booter: Starting SL event loop\n\n");
+	sl_sched_loop();
+}
+
+static int
+boot_check_scheduler(char *comp_name)
```

Comment:
This hack is gross, but probably fine for now. We should fix this/put this information in the bootscript in the future.